Data source vault auth backends

mengesb opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-vault/blob/master/.github/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #1793
Closes #1081 

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
FEATURES:
* Add new data source `vault_auth_backends`, test fixup by @fairclothjm
```

Output from acceptance testing:

```text
$ vault server -dev >vault-server-dev.log 2>&1 &
$ export TESTARGS="--run TestDataSourceAuthBackends"
$ export VAULT_ADDR='http://127.0.0.1:8200'
$ export VAULT_TOKEN=REDACTED
$ GOPATH=$CWD make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestDataSourceAuthBackends -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template(cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.377s
```

fairclothjm commented on vault/data_source_auth_backends_test.go:
```diff
@@ -0,0 +1,96 @@
+}
+data "vault_auth_backends" "test" {
+	depends_on = [
+		"vault_auth_backend.test-foo",
+		"vault_auth_backend.test-bar",
```
I know it is a test but Quoted references are deprecated and may be removed in future versions so we should probably remove them.
```suggestion
		vault_auth_backend.test-foo,
		vault_auth_backend.test-bar,
```

fairclothjm commented on vault/data_source_auth_backends.go:
```diff
@@ -0,0 +1,80 @@
+	d.Set("type", targetType)
+	d.Set("accessors", accessors)
+
+	// If we fell out here then we didn't find our Auth in the list.
```
I don't think this comment is necessarily true. Can we remove it?

mengesb replied:
This was adopted from `data.auth_backend`; I'll remove it

fairclothjm commented on website/docs/d/auth_backends.html.md:
```diff
@@ -0,0 +1,39 @@
+  The `namespace` is always relative to the provider's configured [namespace](/docs/providers/vault#namespace).
+  *Available only for Vault Enterprise*.
+
+* `type` - (Optional) The name of the auth method type.
```
This describes what this field is but could we provide a little information on what this does?

fairclothjm replied:
Yes

mengesb replied:
I added a few more words - LMK if this is to your liking ;)

fairclothjm commented on vault/data_source_auth_backends.go:
```diff
@@ -0,0 +1,80 @@
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "The auth backend mount points.",
+			},
+			"type": {
```
Can we use consts for each of these? If not already defined in https://github.com/hashicorp/terraform-provider-vault/blob/main/internal/consts/consts.go then we can add new entries.
```suggestion
			consts.FieldType: {
```

fairclothjm commented on vault/data_source_auth_backends.go:
```diff
@@ -0,0 +1,80 @@
+		if targetType == "" {
+			paths = append(paths, path)
+			accessors = append(accessors, auth.Accessor)
+			// we do this to make test assertions easier
+			// this is not required
+			sort.Strings(paths)
```
This doesn't seem to be required for the tests to pass. I don't know why I added it 🤔 . Can we remove it?

mengesb replied:
I'll remove it and re-run the tests to be sure. The tests depended on an order, not sure if that will upset tests; will verify

mengesb replied:
```text
--- FAIL: TestDataSourceAuthBackends (2.04s)
    data_source_auth_backends_test.go:22: Step 2/3 error: Check failed: Check 2/6 error: data.vault_auth_backends.test: Attribute 'paths.0' expected "approle", got "userpass"
FAIL
FAIL    github.com/hashicorp/terraform-provider-vault/vault     2.385s
FAIL
make: *** [testacc] Error 1
```

So I can update comparison order, however is it possible that the order would change based on the testing conditions I've supplied?

mengesb replied:
So the first time I run the test, it fails, but subsequent runs pass successfully (with the sort removed)

mengesb replied:
Re-added the sort to ensure tests are consistent. Unless you think it's better to sort only in the test, I'll leave this as-is for now. LMK if you think this is OK and resolved

fairclothjm replied:
I think it is better to sort only in the test. But if we can't do that then I don't see a problem with leaving this as-is. I will leave it up to you.

fairclothjm replied:
I think this is probably better since the sort operation isn't required in the source. Only thing I would suggest is to add a comment on why you are sorting in the test. Thanks!

mengesb replied:
OK, getting this pushed in

fairclothjm commented on vault/data_source_auth_backends_test.go:
```diff
@@ -0,0 +1,96 @@
+}
+data "vault_auth_backends" "test" {
+	depends_on = [
+		"vault_auth_backend.userpass",
+		"vault_auth_backend.approle",
```
I know it is a test but Quoted references are deprecated and may be removed in future versions so we should probably remove them.

mengesb replied:
This should now be resolved

fairclothjm reviewed (commented):
Thanks @mengesb! Just a few small changes and this should be good to go!

fairclothjm commented:
@mengesb RE: consts

Yes, let's update all fields to use consts. If not already defined in https://github.com/hashicorp/terraform-provider-vault/blob/main/internal/consts/consts.go then we can add new entries.

> data.auth_backend doesn't make use of consts

We are requiring that all new resources use consts.

mengesb commented:
```text
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestDataSourceAuthBackends -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role      (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template  (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation     (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.127s
```

mengesb commented:
Myriad of comments, commits, and overall byte sized goodness.

fairclothjm commented on vault/data_source_auth_backends.go:
```diff
@@ -0,0 +1,80 @@
+				Description: "The type of the auth backend.",
+			},
+			consts.FieldAuthMethodAccessors: {
+				//"accessors": {
```
```suggestion
```

fairclothjm commented on vault/data_source_auth_backends.go:
```diff
@@ -0,0 +1,80 @@
+				Optional:    true,
+				Description: "The type of the auth backend.",
+			},
+			consts.FieldAuthMethodAccessors: {
```
I prefer the name `accessors` since this name stutters:

`data.vault_auth_backends.test.auth_method_accessors`

vs

`data.vault_auth_backends.test.accessors`

Also the existing `vault_auth_backend` resource uses `accessor` so I think it makes sense to be similar to that.

fairclothjm commented on vault/data_source_auth_backends_test.go:
```diff
@@ -0,0 +1,96 @@
+				Config: testDataSourceAuthBackendsBasic,
+				// The token auth method is built-in and automatically enabled
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(ds, "paths.#", "1"),
+					resource.TestCheckResourceAttr(ds, "paths.0", "token"),
+					resource.TestCheckResourceAttr(ds, "auth_method_accessors.#", "1"),
```
Can we use consts for all the field references?

mengesb commented:
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestDataSourceAuthBackends -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet(cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template(cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation  (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.140s
```

mengesb commented:
OK

I updated the tests to uses `consts`, and also found my IDE warned me about a duplicate use of `terraform-plugin-sdk/v2/helper/resource` so I removed the renamed import and updated those references. I also added the new const `FieldAccessors` since there was no other plural `accessor` const.

fairclothjm reviewed (commented):
One last nitpick and this PR can be merged

mengesb commented:
```text
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test --run TestDataSourceAuthBackends -timeout 30m ./...
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode0.376s [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/internal/consts   [no test files]
?       github.com/hashicorp/terraform-provider-vault/helper    [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/group   [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/identity/mfa     [no test files]
?       github.com/hashicorp/terraform-provider-vault/internal/pki      [no test files]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode0.686s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet0.987s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role   1.305s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template1.593s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation  1.401s [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/identity/entity  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/internal/provider (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/testutil  (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
ok      github.com/hashicorp/terraform-provider-vault/vault     3.175s
```

mengesb commented:
OK, we're all `consts`-ed; sort from data source removed so there's no implied order, and the `type` field documentation was amended to help explain it a little better. We bounced around on `accessors` due to `consts`, but we're back with `accessors` being a `consts` so I think we're good there.

Hope everything is looking 👍

fairclothjm reviewed (approved):
LGTM! Thanks @mengesb !
